forum/src/main/java/unipd/se18/ocrcamera/forum/viewmodels/AddPost_VM.java

```diff
@@ -0,0 +1,320 @@
+        }
+    };
+
+    /**
+     * Adds a post to the forum
+     * More details at: {@link RequestManager},
+     * {@link RequestManager#sendRequest(Context, ArrayList)}.
+     * @param context The reference of the activity/fragment that calls this method
+     * @param title The new post's title
+     * @param message The new post's message
+     * @param author The new post's author
+     * @author Pietro Prandini (g2)
+     */
+    @Override
+    public void addPostToForum(final Context context, String title, String message, String author) {
+        Log.i(TAG,"addPostToForum");
+
+        // Checks the validity of the parameters
+        if(!areParametersValid(title, message, author)) {
+            Log.d(TAG,"addPostToForum -> The parameters are not valid");
+
+            // The parameters are not valid
+	    // Notifies the event to the View
+            notifier.onNotValidParameters(context.getString(R.string.not_valid_parameters));
+
+            // Ends the method
+            return;
+        }
+
+        // Sets up the manager useful for adding posts
+        RequestManager postManager = new RequestManager();
+
+        // Sets up the manager worker listener
+        postManager.setOnRequestFinishedListener(requestManagerListener);
+
+        /*
+        The next try catch handles a possible failure on the creation
+        of the JSON object that represents a new post.
+         */
+
+        try {
+            // Sets up the parameters for the adding post request
+            ArrayList<RequestManager.Parameter> postManagerParameters =
+                    getAddPostParameters(title, message, author);
+
+            // Sends the request
+            Log.i(TAG,"addPostToForum -> Sending the request");
+            postManager.sendRequest(context, postManagerParameters);
+        } catch (JSONException e) {
+            // JSON Post creation failed
+            e.printStackTrace();
+            // Notifies the error
+            Log.d(TAG, "addPostToForum -> JSON creation problem");
+	    // Notifies the event to the View
+            notifier.onJSONPostCreationFailed(e.toString());
+        }
+    }
+
+    /**
+     * Checks the validity of the parameters received (not null and not empty String)
+     * @param title The new post's title
+     * @param message The new post's message
+     * @param author The new post's author
+     * @return TRUE if the parameters are valid (not null and not empty String), FALSE otherwise
+     * @author Pietro Prandini (g2)
+     */
+    private Boolean areParametersValid(String title, String message, String author) {
+        // Checks if the Strings are not null and not empty ones
+        return (title != null && !title.equals(""))
+                && (message != null && !message.equals(""))
+                && (author != null && !author.equals(""));
+    }
+
+    /*
+    The next two methods are public and static to favourite the testing
```

Comment:
Yes, for testing i've meant junit tests and instrumented tests of android (in the instrumented test I've used some mocked object but it's not relevant) as I've written in the description of this pull request. Should I write more details in this comment?

The story about that is this: I'd to respect an interface that has imposed to me the parameters and the return type of the method addPostToForum but at the same time I'd to split that method in plus that one part for granting a certain level of readability.
So I'd use the next two method for preparing the request (an arrylist of parameters) and for preparing the String of the post in a JSON format.
If these two methods would be private I wouldn't had accessed to them for testing them but I've thought that was important test to avoid the risk to send something wrong to the server and in this sense I've preferred to test these method. But these two methods only prepare a request (ones return a JSON string and ones return an arraylist of parameters), the real senders is addPostToForum.

Reply:
Ok, the code is already commented enough, no need to add more details.
Regardless I have now understood more clearly what you meant to do. Thank you